### DISCUSSION TITLE
How to explode a dask dataframe column horizontally instead of vertically

### DISCUSSION THREAD

**Discussion by damimavencode** [dask/dask#11561] [Q&A]
Thanks for the response.

I am trying to explode the column of a Dask dataframe horizontally instead of vertically.

I also want a solution that will work if the length of the list values is arbitrary or not known in advance as that could impact how the meta parameter is constructed

Please note that all list entries in a column will have the same length. However, this length can vary and cannot always be deterministic in advance.

This is the current Pandas dataframe that works

Pandas DataFrame Approach

import pandas as pd

data = {‘list_column’: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]}

df = pd.DataFrame(data)

df_exploded = df.explode(‘list_column’)

df_horizontal = df_exploded.pivot_table(index=df_exploded.index, columns=df_exploded.groupby(df_exploded.index).cumcount(), values=‘list_column’)

df_horizontal

However, from the pure Dask Dataframe point of view, this is what I am actually looking for which I have not been able to get to work as I am always getting a ValueError: Grouper and axis must be same length

Dask DataFrame Approach

import pandas as pd
import dask.dataframe as dd

data = {‘list_column’: [[1, 2, 3], [4, 5, 6], [7, 8, 9]]}

pdf = pd.DataFrame(data)

ddf = dd.from_pandas(pdf, npartitions=1)

ddf_exploded = ddf.explode(‘list_column’)

ddf_horizontal = ddf_exploded.pivot_table(index=ddf_exploded.index, columns=ddf_exploded.groupby(df_exploded.index).cumcount(), values=‘list_column’)

ddf_horizontal.compute()

Thank you so much for your help in this regard.

**Comment by Samielakkad** [comment]
Dask needs to know the output columns (the `meta`) when it builds the task graph, so a truly unknown number of columns cannot be introduced lazily inside `map_partitions`. A practical solution is to compute only the small scalar that describes the width, then build the real lazy computation with explicit metadata.

```python
import pandas as pd

n = int(
    ddf["list_column"]
    .dropna()
    .map(len, meta=("length", "int64"))
    .max()
    .compute()
)

def expand_partition(pdf, width):
    out = pd.DataFrame(pdf["list_column"].tolist(), index=pdf.index)
    out = out.reindex(columns=range(width))
    out.columns = [f"item_{i}" for i in range(width)]
    return out

meta = pd.DataFrame({
    f"item_{i}": pd.Series(dtype="object") for i in range(n)
})

wide = ddf.map_partitions(expand_partition, n, meta=meta)
```

Use the actual element dtype instead of `object` if it is known. This performs one eager reduction to discover the schema, but the large expansion remains partitioned and lazy.

The original approach also appears to mix a Pandas index (`df_exploded.index`) into a Dask expression. Avoid that, because it can cause alignment and graph-construction problems. A pivot is not a good fit here either: it also needs a known set of output columns. `map_partitions` with explicit `meta` is the normal Dask pattern for this shape change.
